### PR TITLE
OCPBUGS-1627: [vsphere-zones] Fix user folders

### PR DESCRIPTION
**Pull request by jcpowermac**
- Adds length check for folderPath split
- Only create a folder when folder name is empty
- Add additional validation for resource pools and folders
- In the vsphereprivate terraform provider support
the ability to use folders as full paths
- In terraform trimprefix of the folder path based on
`/<datacenter>/vm`

### CODE REVIEW

File: pkg/tfvars/vsphere/vsphere.go [hunk 1]
```diff
@@ -131,11 +145,18 @@ func createDatacenterFolderMap(infraID string, failureDomains []vtypes.FailureDo
 			tempFolder.Name = infraID
 			failureDomains[i].Topology.Folder = infraID
 		}
+		folder, err := finder.Folder(context.TODO(), tempFolder.Name)
 
-		key := fmt.Sprintf("%s-%s", tempFolder.Datacenter, tempFolder.Name)
-		folders[key] = tempFolder
+		if err != nil {
+			return nil, err
+		}
```
Comment (bostrt) [inline pkg/tfvars/vsphere/vsphere.go:152]: I think `finder.Folder` will return an error when no folder is found so you may need to check for that here

File: pkg/tfvars/vsphere/vsphere.go [hunk 2]
```diff
@@ -117,11 +124,18 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
 // The datacenter could be reused but a folder could be
 // unique - the key then becomes a string that contains
 // both the datacenter name and the folder to be created.
-func createDatacenterFolderMap(infraID string, failureDomains []vtypes.FailureDomain) map[string]*folder {
+func createDatacenterFolderMap(infraID string, failureDomains []vtypes.FailureDomain, vcenters []vtypes.VCenter) (map[string]*folder, error) {
 	folders := make(map[string]*folder)
+	soap, _, cleanup, err := vsphere.CreateVSphereClients(context.TODO(), vcenters[0].Server, vcenters[0].Username, vcenters[0].Password)
```
Comment (bostrt) [inline pkg/tfvars/vsphere/vsphere.go:129]: `vcenters` may be zero length 

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/6412/pull-ci-openshift-installer-master-e2e-vsphere-ovn/1574388049691807744

**Comment by jinyunma** [comment]
tested on pre-merged payload 4.12.0-0.ci.test-2022-09-27-002038-ci-ln-lbsmvtb-latest
when platform.vsphere.failureDomains.topology.folder set as folder name without path, `folder: qe-jima-ibmcloud`, get panic error, should it be absolute path here? current platform.vsphere.folder requires absolute path. 
```
INFO The file was found in cache: /home/user/.cache/openshift-installer/image_cache/rhcos-412.86.202208101039-0-vmware.x86_64.ova. Reusing... 
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
github.com/openshift/installer/pkg/tfvars/vsphere.TFVars({{0xc00139b7b8, 0x3, 0x3}, {0xc0011668b8, 0x12}, {0xc0011668a0, 0x14}, {0xc0011667f8, 0x14}, {0xc0012e85a0, ...}, ...})
	/go/src/github.com/openshift/installer/pkg/tfvars/vsphere/vsphere.go:82 +0x65b
github.com/openshift/installer/pkg/asset/cluster.(*TerraformVariables).Generate(0x1d46e820, 0x5?)
	/go/src/github.com/openshift/installer/pkg/asset/cluster/tfvars.go:847 +0x4798
```
when unset `platform.vsphere.vcenters`, set `platform.vsphere.folder` as `"/IBMCloud/vm/qe-jima-ibmcloud"`, and no folder setting in topology of failureDomains. I see that failureDomain reuse `platform.vsphere.folder`, but terraform fails when importing ova image on failureDomain in datacneter-2 .
```
time="2022-09-27T06:10:44Z" level=debug msg="vsphereprivate_import_ova.import[1]: Creation complete after 2m22s [id=vm-373458]"
time="2022-09-27T06:10:44Z" level=error
time="2022-09-27T06:10:44Z" level=error msg="Error: failed to find provided vSphere objects: folder '/datacenter-2/vm//IBMCloud/vm/qe-jima-ibmcloud' not found"
time="2022-09-27T06:10:44Z" level=error
time="2022-09-27T06:10:44Z" level=error msg="  with vsphereprivate_import_ova.import[3],"
time="2022-09-27T06:10:44Z" level=error msg="  on main.tf line 70, in resource \"vsphereprivate_import_ova\" \"import\":"
time="2022-09-27T06:10:44Z" level=error msg="  70: resource \"vsphereprivate_import_ova\" \"import\" {"
time="2022-09-27T06:10:44Z" level=error
time="2022-09-27T06:10:44Z" level=error msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failure applying terraform for \"pre-bootstrap\" stage: failed to create cluster: failed to apply Terraform: exit status 1\n\nError: failed to find provided vSphere objects: folder '/datacenter-2/vm//IBMCloud/vm/qe-jima-ibmcloud' not found\n\n  with vsphereprivate_import_ova.import[3],\n  on main.tf line 70, in resource \"vsphereprivate_import_ova\" \"import\":\n  70: resource \"vsphereprivate_import_ova\" \"import\" {\n\n"
```

**Comment by jcpowermac** [comment]
> when platform.vsphere.failureDomains.topology.folder set as folder name without path, folder: qe-jima-ibmcloud, get panic error, should it be absolute path here? current platform.vsphere.folder requires absolute path.

@rvanderp3 isn't there validation for absoulte path?

_weird_ it panic'ed here
https://github.com/openshift/installer/pull/6412/files#diff-cdefcb6fa878811aa273941a4ac54baf6ce971ab287ef81f2736a9d9904cbd61R82

> I see that failureDomain reuse platform.vsphere.folder, but terraform fails when importing ova image on failureDomain in datacneter-2 .

That's certainly not right.


@jinyunma just to make it easier on us, can you add the two install-configs on the bug?

File: pkg/types/vsphere/validation/platform.go [hunk 3]
```diff
@@ -85,18 +85,26 @@ func validateMultiZone(p *vsphere.Platform, fldPath *field.Path) field.ErrorList
...
+
+			/*  Since these two parameters are optional, not sure you can
+			replace these values...
+			what if the datacenter in the failure domain is not the
+			same as what is defined in either p.ResourcePool or p.Folder?
+
+			if len(failureDomain.Topology.ResourcePool) == 0 {
+				p.FailureDomains[idx].Topology.ResourcePool = p.ResourcePool
+			}
 			if len(failureDomain.Topology.Folder) == 0 {
```
Comment (jcpowermac) [inline pkg/types/vsphere/validation/platform.go:103]: If folder doesn't live in datacenter...error and state to specify topology

File: pkg/tfvars/vsphere/vsphere.go [hunk 4]
```diff
@@ -117,25 +124,48 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
 // The datacenter could be reused but a folder could be
 // unique - the key then becomes a string that contains
 // both the datacenter name and the folder to be created.
-func createDatacenterFolderMap(infraID string, failureDomains []vtypes.FailureDomain) map[string]*folder {
+func createDatacenterFolderMap(infraID string, failureDomains []vtypes.FailureDomain, vcenters []vtypes.VCenter) (map[string]*folder, error) {
```
Comment (jcpowermac) [inline pkg/tfvars/vsphere/vsphere.go:127]: revert to existing folder map create

File: pkg/types/vsphere/validation/platform.go [hunk 5]
```diff
@@ -85,18 +85,26 @@ func validateMultiZone(p *vsphere.Platform, fldPath *field.Path) field.ErrorList
 			if len(failureDomain.Topology.ComputeCluster) == 0 {
 				p.FailureDomains[idx].Topology.ComputeCluster = fmt.Sprintf("/%s/host/%s", p.Datacenter, p.Cluster)
 			}
-			if len(failureDomain.Topology.ResourcePool) == 0 {
```
Comment (jcpowermac) [inline pkg/types/vsphere/validation/platform.go:88]: same as folder

**Comment by jinyunma** [comment]
tested on pre-merged payload 4.12.0-0.ci.test-2022-10-19-012832-ci-ln-fm1rvx2-latest and passed. Details can be found in bug comments.
/label qe-approved

**Comment by sadasu** [comment]
Do we need any additional unit tests to exercise validation code with incorrect "folder"? I see some changes to the unit test code but those seem to be to related to updates for the happy path.
If we always expect "folder" to be a path with "/vm/" substring in it, should the minlength be 1? Also, would if we always expect "/vm/" to be part of the path, should be use a kubebuilder validation pattern to make sure that is the case?

**Comment by sadasu** [comment]
Do we need any additional unit tests to exercise validation code with incorrect "folder"? I see some changes to the unit test code but those seem to be to related to updates for the happy path.
If we always expect "folder" to be a path with "/vm/" substring in it, should the minlength be 1? Also, would if we always expect "/vm/" to be part of the path, should be use a kubebuilder validation pattern to make sure that is the case? I see a similar check  in https://github.com/openshift/installer/pull/6412/files#diff-daaac9b1d9031588d68fc619ca96e45f98e8287293ea2cd157fa8ad93915e91dR126.

**Review by patrickdillon — COMMENTED** [review]
/approve

In general this LGTM. Would be a little easier to follow if the commits were broken up, but probably not worth it at this stage.

**Comment by jcpowermac** [comment]
/retest-required

@patrickdillon any chance I could get a skip here?

**Pull request merged**